File: kangtaku/binary_tree_right_side_view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans=new ArrayList<>();
        find(ans,root,0);
        return ans;
    }
    public void find(List<Integer> ans,TreeNode root,int depth){
        if(root==null)return;
        if(ans.size()==depth)ans.add(0);
        ans.set(depth,root.val);
        find(ans,root.left,depth+1);
        find(ans,root.right,depth+1);
    }
}

File: iamminji/leetcode/medium/binary_tree_right_side_view.py
# 199. Binary Tree Right Side View
# https://leetcode.com/problems/binary-tree-right-side-view/description/


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def rightSideView(self, root):
        cache = dict()
        max_depth = -1

        stack = [(root, 0)]
        while stack:
            node, depth = stack.pop()

            if node is not None:
                max_depth = max(max_depth, depth)

                if depth not in cache:
                    cache[depth] = node.val

                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))

        return [cache[depth] for depth in range(max_depth + 1)]
